diff --git a/node_modules/auth-astro/src/integration.ts b/node_modules/auth-astro/src/integration.ts
index f749024..9200d5c 100644
--- a/node_modules/auth-astro/src/integration.ts
+++ b/node_modules/auth-astro/src/integration.ts
@@ -41,7 +41,7 @@ export default (config: AstroAuthConfig = {}): AstroIntegration => ({
 				logger.error('No Adapter found, please make sure you provide one in your Astro config')
 			}
 			const edge = ['@astrojs/vercel/edge', '@astrojs/cloudflare'].includes(
-				astroConfig.adapter.name
+				astroConfig.adapter!.name
 			)
 
 			if (
